### PR TITLE
[Jormungandr] Harmonize multilanguage parameters in navitia

### PR DESCRIPTION
**Pull request by kadhikari**
The followings are some details:
- Only one parameter **language** should be used instead of three existing parameters.
- Use of 11 languages allowed in navitia but all languages are not used in each **street_network** connector.
- In navitia, use of wrong value raises error with a message = **Select a specific language for street network instructions ...**
- if no `language` is provided in request, using the value configured in the coverage (default value for coverage = fr-FR)
- then if the language chosen in navitia is absent in a street_network connector, en-US will be used.
- never supposed to happen, but to avoid crash: in all connectors final fallback is to `fr-FR`

Note: The existing attribute asgard_language should be deleted once the next release is deployed to all the platforms.

For more details:  https://navitia.atlassian.net/browse/NAV-2152

### CODE REVIEW

File: source/jormungandr/jormungandr/street_network/asgard.py [hunk 1]
```diff
@@ -54,35 +54,16 @@
...
-    german = "de-DE"
-    greek = "el-GR"
-    english_gb = "en-GB"
-    english_pirate = "en-US-x-pirate"
-    english_us = "en-US"
-    spanish = "es-ES"
-    estonian = "et-EE"
-    finnish = "fi-FI"
+    dutch = "nl-NL"
+    english = "en-GB"
```
Comment (woshilapin) [inline source/jormungandr/jormungandr/street_network/asgard.py]: Maybe `en-US` is a more sensible default? It's usually the default in a lot of systems, and we usually tend to speak/translate to a US variant, more than to a GB variant.

Note that if you agree to use `en-US` as a default value, there might a lot of places in your PR to modify too (in order to make it the default everywhere).

File: source/jormungandr/jormungandr/street_network/here.py [hunk 2]
```diff
@@ -66,22 +66,18 @@
...
+
 # Possible values implemented. Full languages within the doc:
 # https://developer.here.com/documentation/routing/dev_guide/topics/resource-param-type-languages.html#languages
 # Be careful, the syntax has to be exact
 class Languages(Enum):
-    afrikaans = "af"
-    arabic = "ar-sa"
-    chinese = "zh-cn"
     dutch = "nl-nl"
     english = "en-gb"
```
Comment (woshilapin) [inline source/jormungandr/jormungandr/street_network/here.py]: I made a comment related to a default value being `en-US` instead of `en-GB` for Asgard. If you decide to change the default value in Asgard, you probably need to modify this one too for coherence.

File: source/jormungandr/jormungandr/street_network/here.py [hunk 3]
```diff
@@ -400,11 +396,8 @@ def _get_language(self, language):
             return Languages.english
 
     def get_language_parameter(self, request):
-        _language = request.get('_here_language', None)
-        if _language == None:
-            return self.language
-        else:
-            return self._get_language(_language.lower())
+        language = request.get('language', None)
```
Comment (woshilapin) [inline source/jormungandr/jormungandr/street_network/here.py]: Should this default to `english`?
```suggestion
        language = request.get('language', "english")
```

File: source/jormungandr/jormungandr/street_network/tests/handimap_test.py [hunk 4]
```diff
@@ -260,7 +260,7 @@ def get_language_handimap_func_language_invalid_test():
     instance = MagicMock()
     handimap = Handimap(instance=instance, service_url=fake_service_url)
     language = handimap._get_language("toto")
-    assert language == "fr-FR"
+    assert language == "en-EN"
```
Comment (woshilapin) [inline source/jormungandr/jormungandr/street_network/tests/handimap_test.py]: `en-EN` doesn't exist. See [here](https://en.wikipedia.org/wiki/Language_localisation#Language_tags_and_codes) for an exhaustive list of existing variants.

File: source/tyr/tyr/resources.py [hunk 5]
```diff
@@ -882,72 +882,34 @@ def put(self, version=0, id=None, name=None):
...
+            help='Select a specific language for street network instructions.\n'
             'list available:\n'
-            '- bulgarian = bg-BG\n'
-            '- catalan = ca-ES\n'
-            '- czech = cs-CZ\n'
-            '- danish = da-DK\n'
-            '- german = de-DE\n'
-            '- greek = el-GR\n'
+            '- dutch = nl-NL\n'
             '- english_gb = en-GB\n'
```
Comment (woshilapin) [inline source/tyr/tyr/resources.py]: Should this be edited to remove the variant?

```suggestion
            '- english = en-GB\n'
```

**Review by SGrenet — COMMENTED** [review]
need navitia-configurator PR too to change asgard_language to language
https://github.com/hove-io/navitia-configurator/blob/fbf8cb3e8b3e0691a03b122c1cdc5fd75c099642/coverages.py#L86

and how to switch ? hard manual sync or smooth with retro-compatibility?

File: source/jormungandr/jormungandr/street_network/here.py (hunk 3 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/street_network/here.py]: :+1:

File: source/jormungandr/jormungandr/street_network/tests/handimap_test.py (hunk 4 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/street_network/tests/handimap_test.py]: Since list of languages managed by handimap is not modified, i didn't even test if transformation of english to en-EN is valid or not

File: source/tyr/tyr/resources.py (hunk 5 above)
Comment (kadhikari) [reply source/tyr/tyr/resources.py]: :+1:

File: source/jormungandr/jormungandr/street_network/tests/handimap_test.py (hunk 4 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/street_network/tests/handimap_test.py]: I verified in the deo handimap:
https://dev.keolis-rennes.handimap.fr/docs/#model-RouteRequest
language	Languagesstring
default: en-EN
example: fr-FR

File: source/jormungandr/jormungandr/street_network/asgard.py (hunk 1 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/street_network/asgard.py]: :+1:

File: source/jormungandr/jormungandr/street_network/here.py (hunk 2 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/street_network/here.py]: :+1:

**Comment by kadhikari** [comment]
> need navitia-configurator PR too to change asgard_language to language https://github.com/hove-io/navitia-configurator/blob/fbf8cb3e8b3e0691a03b122c1cdc5fd75c099642/coverages.py#L86
> 
> and how to switch ? hard manual sync or smooth with retro-compatibility?

The new parameter language in navitia as well as attribute in the database has a default value 'french'.
we don't need to modify in navitia-configurator for the existing configuration  asgard_language: str = "french"
Will have to update la config after deploying navitia with future release.

File: source/jormungandr/jormungandr/street_network/tests/handimap_test.py (hunk 4 above)
Comment (jthiard) [reply source/jormungandr/jormungandr/street_network/tests/handimap_test.py]: Hi @woshilapin and @kadhikari,  Handimap developer here.
This is definitely a documentation error on our side. `en-EN` indeed does not exists in ISO language tags and will fallback on our side to `en-US`, this is why it kinda work despite our documentation mistake.
You can safely change to `en-US`, I'll update the docs.
Sorry for that.

Comment (woshilapin) [reply source/jormungandr/jormungandr/street_network/tests/handimap_test.py]: Thank you very much for showing up and posting this :smile:

File: source/jormungandr/jormungandr/street_network/handimap.py [hunk 6]
```diff
@@ -54,10 +54,7 @@
     'url': 'https://www.handimap.fr',
 }
 
-
-class Languages(Enum):
-    french = "fr-FR"
-    english = "en-EN"
+LANGUAGE_TRANSFORMATION_LIST = {"en-US": "en-EN", "en-GB": "en-EN", "fr-FR": "fr-FR"}
```
Comment (jthiard) [inline source/jormungandr/jormungandr/street_network/handimap.py]: As stated in another comment, this "en-EN" language tag was a documentation error on our side (Handimap).
You can safely replace it with a more conventional `en-US` without any risk (`en-US` is our default, and already used used as fallback for any unknown language tag like `en-EN`).
Sorry for that documentation fail.

File: source/jormungandr/jormungandr/street_network/tests/handimap_test.py (hunk 4 above)
Comment (jthiard) [reply source/jormungandr/jormungandr/street_network/tests/handimap_test.py]: You're welcome, feel free to tag me for any issue with Handimap API.
I just updated the docs at https://dev.keolis-rennes.handimap.fr/docs

File: source/jormungandr/jormungandr/street_network/handimap.py (hunk 6 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/street_network/handimap.py]: > As stated in another comment, this "en-EN" language tag was a documentation error on our side (Handimap). You can safely replace it with a more conventional `en-US` without any risk (`en-US` is our default, and already used used as fallback for any unknown language tag like `en-EN`). Sorry for that documentation fail.

Hi @jthiard will it be possible to put the link of the DOC Hadimap as I don't have other than  https://dev.keolis-rennes.handimap.fr/docs/ and it give mes [ fr-FR, en-EN ] as language ?
Thanks

Comment (jthiard) [reply source/jormungandr/jormungandr/street_network/handimap.py]: Hi @kadhikari sure, we have two urls : 

-  https://dev.keolis-rennes.handimap.fr/docs/ : our staging instance
-  https://keolis-rennes.handimap.fr/docs/ our prod instance

At this time the two use the same version since there is no future version in staging yet.

If you still see `en-EN` on https://dev.keolis-rennes.handimap.fr/docs/ it may be your browser cache. You can try https://keolis-rennes.handimap.fr/docs/ which may be not cached on your side.

I just changed the webserver config to send "no cache" headers for our openAPI file to avoid this cache problem in the future.

File: source/jormungandr/jormungandr/street_network/here.py [hunk 7]
```diff
@@ -392,19 +389,16 @@ def get_exclusion_areas(self, request):
...
+        return language
 
     def get_language_parameter(self, request):
-        _language = request.get('_here_language', None)
-        if _language == None:
-            return self.language
-        else:
-            return self._get_language(_language.lower())
+        language = request.get('language', 'fr-FR')
+        return self.language if not language else self._get_language(language)
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/here.py]: It looks like you deal with the default value in too many places, so it ends up being inconsistent between fr-Fr en en-US.
I'm suggesting using `None` most of the time, and default in `None` case only at the end, once and for all.

Here, the second line is useless if you have a default in the first one, I think:
```suggestion
        language = request.get('language', None)
        return self.language if not language else self._get_language(language)
```

File: source/jormungandr/jormungandr/street_network/tests/asgard_test.py [hunk 8]
```diff
@@ -44,7 +44,7 @@ class FakeInstance(object):
 
     name = 'fake_instance'
     context = None
-    asgard_language = "en-US"
+    language = "en-GB"
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/tests/asgard_test.py:47]: Out of curiousity only: Why did you change that in this test?

File: source/jormungandr/jormungandr/street_network/tests/handimap_test.py [hunk 9]
```diff
@@ -273,15 +273,15 @@ def get_language_handimap_func_language_valid_test():
...
     assert language == "fr-FR"
 
 
 def get_language_parameter_handimap_func_language_invalid_test():
     instance = MagicMock()
     handimap = Handimap(instance=instance, service_url=fake_service_url)
-    request = {"_handimap_language": "english"}
+    request = {"language": "english"}
     language = handimap.get_language_parameter(request)
     assert language == "en-EN"
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/tests/handimap_test.py]: if "english" is invalid (like "toto"), how can you have a different result?

File: source/jormungandr/jormungandr/street_network/handimap.py [hunk 10]
```diff
@@ -85,7 +82,7 @@ def __init__(
         self.headers = {"Content-Type": "application/json", "Accept": "application/json"}
         self.timeout = timeout
         self.modes = modes if modes else ["walking"]
-        self.language = self._get_language(kwargs.get('language', "french"))
+        self.language = self._get_language(kwargs.get('language', "fr-FR"))
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/handimap.py]: Shouldn't you default to `None` here, and then in `_get_language` use the one from the instance in case of `None`?

File: source/jormungandr/jormungandr/street_network/handimap.py (hunk 6 above)
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/handimap.py]: ~After previous review-conversation, it looks like this is not a transformation map anymore, just a validity list.~
Switch to en-US instead of en-EN

File: source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py [hunk 11]
```diff
@@ -585,7 +585,7 @@ def _build_fallback(
     car_park_crowfly_duration = None
     via_pt_access = None
     via_poi_access = None
-    language = request.get('_asgard_language', "english_us")
+    language = request.get('language', "en-US")
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py]: Same for https://github.com/hove-io/navitia/pull/4084/files#diff-1db49300dbe77e5d5ef58deb0ebd54793fd7851c5f9a3b238ebd4091567bf5caR133 and https://github.com/hove-io/navitia/pull/4084/files#diff-3307dc8a614b49809e0d0843ff12386f70537d2f093790b1d2872d6fd89ab70eR352:
Here you don't "respect" the default value of the coverage, no?
I think you should default to None and have asgard-connector deal with the default(coverage) value in a single place.

File: source/jormungandr/jormungandr/street_network/asgard.py [hunk 12]
```diff
@@ -218,9 +183,7 @@ def make_location(self, obj):
         )
 
     def get_language_parameter(self, request):
-        language = request.get('_asgard_language', "english_us")
-        language_tag = getattr(Languages, language, Languages.english_us).value
-        return language_tag
+        return request.get('language', "fr-FR")
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/asgard.py]: Shouldn't you use the coverage's value if language isn't in request?

File: source/jormungandr/jormungandr/interfaces/v1/Journeys.py [hunk 13]
```diff
@@ -528,58 +528,35 @@ def __init__(self):
...
+                    'de-DE',
+                    'hi-IN',
+                    'it-IT',
+                    'ja-JP',
+                    'pt-PT',
+                    'ru-RU',
+                    'es-ES',
                 ]
             ),
             hidden=True,
```
Comment (pbougue) [inline source/jormungandr/jormungandr/interfaces/v1/Journeys.py]: We should unhide it now

File: source/jormungandr/jormungandr/street_network/here.py [hunk 14]
```diff
@@ -392,19 +389,16 @@ def get_exclusion_areas(self, request):
...
-        try:
-            return Languages[language]
-        except KeyError:
-            self.log.error('Here parameters language={} not exist - force to english'.format(language))
-            return Languages.english
+    def _get_language(self, language_value):
+        language = LANGUAGE_TRANSFORMATION_LIST.get(language_value)
+        if not language:
+            self.log.error('Here parameter language={} Invalid - fallback to english'.format(language_value))
+            language = "en-EN"
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/here.py]:
```suggestion
            language = "en-US"
```

File: source/jormungandr/jormungandr/street_network/tests/asgard_test.py (hunk 8 above)
Comment (pbougue) [reply source/jormungandr/jormungandr/street_network/tests/asgard_test.py:47]: As discussed: to check that the value is really handled, and also that it's not one of the default values saving the test.

File: source/jormungandr/jormungandr/street_network/tests/handimap_test.py [hunk 15]
```diff
@@ -273,15 +273,15 @@ def get_language_handimap_func_language_valid_test():
 def get_language_parameter_handimap_func_language_invalid_test():
     instance = MagicMock()
     handimap = Handimap(instance=instance, service_url=fake_service_url)
-    request = {"_handimap_language": "toto"}
+    request = {"language": "toto"}
     language = handimap.get_language_parameter(request)
     assert language == "fr-FR"
 
 
 def get_language_parameter_handimap_func_language_invalid_test():
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/tests/handimap_test.py]: Naming is the same: only one test is triggered

File: source/jormungandr/jormungandr/street_network/here.py (hunk 7 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/street_network/here.py]: :+1:

File: source/jormungandr/jormungandr/street_network/handimap.py (hunk 6 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/street_network/handimap.py]: :+1:

File: source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py (hunk 11 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py]: It's nearly impossible to have language = None, and it arrives default language in navitia before using connector street_network will be **fr-FR**
Since instance is absent here, no way to get instance.language !

File: source/jormungandr/jormungandr/interfaces/v1/Journeys.py (hunk 13 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/interfaces/v1/Journeys.py]: :+1:

File: source/jormungandr/jormungandr/street_network/here.py (hunk 14 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/street_network/here.py]: :+1:

File: source/jormungandr/jormungandr/street_network/tests/handimap_test.py (hunk 15 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/street_network/tests/handimap_test.py]: :+1:

File: source/jormungandr/jormungandr/street_network/asgard.py (hunk 12 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/street_network/asgard.py]: :+1:

File: source/jormungandr/jormungandr/street_network/tests/handimap_test.py (hunk 9 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/street_network/tests/handimap_test.py]: :+1:

File: source/jormungandr/jormungandr/street_network/handimap.py (hunk 10 above)
Comment (kadhikari) [reply source/jormungandr/jormungandr/street_network/handimap.py]: :+1:

File: source/jormungandr/jormungandr/street_network/handimap.py [hunk 16]
```diff
@@ -117,19 +114,15 @@ def status(self):
...
+        language = LANGUAGE_TRANSFORMATION_LIST.get(language_value)
+        if not language:
+            self.log.error('Handimap parameter language={} Invalid - fallback to english'.format(language_value))
+            language = "en-US"
+        return language
 
     def get_language_parameter(self, request):
-        language = request.get('_handimap_language', None)
+        language = request.get('language')
         return self.language if not language else self._get_language(language.lower())
```
Comment (jthiard) [inline source/jormungandr/jormungandr/street_network/handimap.py]: Since I'm still around :sweat_smile: 
I think you  want to remove the `lower()` call here, there is no way a lowered language tag could ever match something in ` LANGUAGE_TRANSFORMATION_LIST`.

Comment (kadhikari) [reply source/jormungandr/jormungandr/street_network/handimap.py]: Thanks a lot

**Review by pbench — APPROVED** [review]
